base/rational.jl

```diff
@@ -129,12 +129,220 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
+    RoundingMode{:Nearest}, RoundingMode{:NearestTiesAway}}
+
+rounding_mode_translated_for_abs(rm::RoundingModesIndependentFromSign, ::Real) = rm
+
+function rounding_mode_translated_for_abs(::RoundingMode{:Down}, sign::Real)
+    !signbit(sign) ? RoundToZero : RoundFromZero
+end
+
+function rounding_mode_translated_for_abs(::RoundingMode{:Up}, sign::Real)
+    !signbit(sign) ? RoundFromZero : RoundToZero
+end
+
+clamped_to_zero(x) = max(zero(x), x)
+
+struct FloatComponentsResult{S<:Integer,T<:Integer}
+    mantissa::S
+    exponent::T
+    inexact::Bool
+    underflowed::Bool
+
+    function FloatComponentsResult(m::S, e::T;
+                                   inexact::Bool,
+                                   underflowed::Bool) where {S<:Integer,T<:Integer}
+        new{S,T}(m, e, inexact, underflowed)
+    end
+end
+
+# `num`, `den` are positive integers. `requested_precision` is the
+# requested floating-point precision. `T` is the integer type that
+# we'll be working with mostly, it needs to be wide enough.
+# `exp_lower_bound` is the minimum allowed normalized exponent for
+# normal numbers.
+function rational_to_float_components_impl(num::Integer,
+                                           den::Integer,
+                                           requested_precision::Integer,
+                                           ::Type{T},
+                                           romo::RoundingMode,
+                                           exp_lower_bound::Union{Nothing,Integer}) where
+{T<:Integer}
+    num_bit_width = bit_width(num)
+    den_bit_width = bit_width(den)
+
+    # `T` must be wide enough to avoid overflow.
+    numT = T(num)
+    denT = T(den)
+
+    # Creates a mantissa in `quo_` that's at least
+    # `requested_precision` bits wide.
+    bit_shift_num_ = clamped_to_zero(den_bit_width - num_bit_width + requested_precision)
+    quo_ = div(numT << bit_shift_num_, den, RoundToZero)
+
+    # Nonnegative. Experiments indicate that, when iterating over all
+    # possible numerators and denominators below some bit width, with
+    # some fixed value for `requested_precision`, the maximal attained
+    # value will be
+    # `max(1, max(num_bit_width, den_bit_width) - requested_precision)`.
+    # So in the worst case we have `requested_precision == 1` and
+    # `excess_precision == max(1, max(num_bit_width, den_bit_width) - 1)`
+    excess_precision = bit_width(quo_) - requested_precision
+
+    # Normalized exponent
+    nexp = bit_width(quo_) - bit_shift_num_ - true
+
+    # Take possible underflow into account: if there's underflow, the
+    # precision needs to be less than requested.
+    adjusted_precision = requested_precision
+    if !isnothing(exp_lower_bound)
+        underflow = clamped_to_zero(exp_lower_bound - nexp)
+        adjusted_precision -= underflow
+    end
+    adjusted_excess_precision = bit_width(quo_) - adjusted_precision
+
+    (adjusted_precision < false) && return FloatComponentsResult(
```

Comment:
```suggestion
    (adjusted_precision < 0) && return FloatComponentsResult(
```

As far as I can tell, this will always come from a call to `precision`, which returns an `Int`. Additionally, these kinds of comparisons with an integer constant are optimized very well in LLVM, so there's nothing to be gained by comparing to `false`.